core/src/test/java/org/infinispan/remoting/RemoteShutdownDuringOperationTest.java

```diff
@@ -0,0 +1,195 @@
+import static org.mockito.Mockito.spy;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+
+import org.assertj.core.api.SoftAssertions;
+import org.infinispan.cache.impl.EncoderCache;
+import org.infinispan.commands.read.GetCacheEntryCommand;
+import org.infinispan.commons.util.concurrent.CompletableFutures;
+import org.infinispan.configuration.cache.CacheMode;
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.context.InvocationContext;
+import org.infinispan.distribution.MagicKey;
+import org.infinispan.expiration.impl.TouchCommand;
+import org.infinispan.interceptors.DDAsyncInterceptor;
+import org.infinispan.test.MultipleCacheManagersTest;
+import org.infinispan.test.TestDataSCIImpl;
+import org.infinispan.test.TestingUtil;
+import org.infinispan.test.fwk.CheckPoint;
+import org.infinispan.topology.ClusterTopologyManager;
+import org.testng.annotations.Test;
+
+@Test(groups = "functional", testName = "remoting.RemoteShutdownDuringOperationTest")
+public class RemoteShutdownDuringOperationTest extends MultipleCacheManagersTest {
+
+   private static final String CACHE_NAME = "distSync";
+
+   private TolerateSuspectOperation operation;
+
+   protected RemoteShutdownDuringOperationTest withOperation(TolerateSuspectOperation operation) {
+      this.operation = operation;
+      return this;
+   }
+
+
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      ConfigurationBuilder cb = getDefaultClusteredCacheConfig(CacheMode.DIST_SYNC, false);
+      int numOwners = 2;
+
+      // Necessary to go to the remote node.
+      if (operation == TolerateSuspectOperation.GET) numOwners = 1;
+
+      cb.clustering().hash().numOwners(numOwners);
+      createClusteredCaches(2, CACHE_NAME, TestDataSCIImpl.INSTANCE, cb);
+   }
+
+   @Test
+   public void testShutdownDuringOperation() throws Exception {
+      EncoderCache<MagicKey, String> c0 = cache(0);
+      EncoderCache<MagicKey, String> c1 = cache(1);
+
+      assertThat(c1.getAdvancedCache().getComponentRegistry().getStatus().isTerminated())
+            .isFalse();
+
+      MagicKey key = new MagicKey("remote-key", c1);
+
+      assertThat(c1.put(key, "value")).isNull();
+      assertThat(c0.get(key)).isEqualTo("value");
+
+      ClusterTopologyManager ctm = c0.getAdvancedCache().getComponentRegistry()
+            .getComponent(ClusterTopologyManager.class);
+
+      try {
+         CheckPoint checkPoint = new CheckPoint();
+         CheckPointInterceptor interceptor = new CheckPointInterceptor(checkPoint);
+         extractInterceptorChain(c1).addInterceptor(interceptor, 1);
+
+         ClusterTopologyManager spyCtm = spy(ctm);
+         CompletableFuture<Void> cf = new CompletableFuture<>();
+         doAnswer(ivk -> {
+            // We delay the handling of the leave request until *after* the command is handled.
```

Comment:
Why not just use the CheckPoint here instead and have it return the future from CheckPoint.future?